pkg/controller/registry/resolver/sat/solve.go

```diff
@@ -0,0 +1,94 @@
+package sat
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/irifrance/gini"
+)
+
+// Identifier values uniquely identify particular Installables within
+// the input to a single call to Solve.
+type Identifier string
+
+// Installable values are the basic unit of problems and solutions
+// understood by this package.
+type Installable interface {
+	// Identifier returns the Identifier that uniquely identifies
+	// this Installable among all other Installables in a given
+	// problem.
+	Identifier() Identifier
+	// Constraints returns the set of constraints that apply to
+	// this Installable.
+	Constraints() []Constraint
+}
+
+var Incomplete = errors.New("cancelled before a solution could be found")
+
+// NotSatisfiable is an error composed of a minimal set of applied
+// constraints that is sufficient to make a solution impossible.
+type NotSatisfiable []AppliedConstraint
+
+func (e NotSatisfiable) Error() string {
+	const msg = "constraints not satisfiable"
+	if len(e) == 0 {
+		return msg
+	}
+	s := make([]string, len(e))
+	for i, a := range e {
+		s[i] = a.String()
+	}
+	return fmt.Sprintf("%s: %s", msg, strings.Join(s, ", "))
+}
+
+const (
+	satisfiable   = 1
+	unsatisfiable = -1
+)
+
+// Solve takes a slice containing all Installables and returns a slice
+// containing only those Installables that were selected for
+// installation. The solution will attempt to minimize the total
+// weight of all selected Installables. If no solution is possible, an
+// error is returned.
+func Solve(installables []Installable) ([]Installable, error) {
+	return SolveWithContext(context.Background(), installables)
```

Comment:
nit: could this take a `context` as an argument instead? and if none use a `context.TODO()`? 

From what I recall `context.Background()` should be called only once in the program, at the beginning as the parent context

Reply:
Sure, TODO may be more appropriate for this, although I think the distinction only exists to clarify intent when reading the source (https://github.com/golang/go/blob/master/src/context/context.go#L199-L202). In production I assume SolveWithContext is always going to be used, but it felt convenient to include this contextless signature for things like tests or utilities.